### PR TITLE
feat: predicting test set classes

### PR DESCRIPTION
**Pull request by stulacy**
Adds the ability to predict the class of an unseen test set.

@llwiggins I've made a start on adding this functionality. Rather than asking the user to separately upload a data frame with labels, I think it makes more sense to ask them to just input the test set text label for the test set. If we want the user to be able to upload multiple test sets then we can code that in just as you've done for the multiple training set groups.

If we only want users to upload a single test set that can contain multiple different classes, then it would be best to require that there is a `label` column in the CSV containing the actual class for each row. However, to be consistent the training data should also take this form of having a single CSV upload with a `label` column.

In my opinion it's more user friendly to allow multiple training and/or test sets as separate CSV files rather than requiring users to merge these outside of the app. I imagine in most use cases users will only run CellPhe over a single timelapse containing the same class. 

If you think this sounds sensible then I can add in the looping ability to upload multiple test sets.

### CODE REVIEW

**Comment by llwiggins** [comment]
Hi Stuart,

This sounds sensible to me, and agree we should avoid anything that requires users to make changes to the CSV file outside of the app! I would imagine the most likely situations are:
(1) User's test set contains just one class and they know the true class label
(2) User's test set is a co-culture containing different classes, but they do not have true class labels

I think your implementation works best for these two cases so I'd be happy for us to go ahead with this!

**Comment by stulacy** [comment]
I've cleaned up the multi-class tab with the agreed change (no longer asking user for a test set label as it's not used in either of the 2 use cases you mentioned) and have added a new plot allowing the user to select a variable to display as a boxplot between the test set's **predicted** classes.

I've also added a ton more changes beyond the original scope of this PR, mostly because it's easier to do it in one place than deal with a merge conflict for each new feature PR.

  - Tweaked layouts, adding inputs directly next to the plots they control
  - Added more text to help make the app more user friendly
  - Changed the density plot in tab 2 so that it's all done in 1 call to `sns.kdeplot`. This results in a ~3x speedup over iterating over each frame and doing it manually
  - In tab 1, selecting the image folder is now down via a file dialog select rather than asking the user to enter the path as a string
  - In tab 1 the input folder is now validated to ensure it contains at least 1 tif

If that all sounds good I can merge it in. It would be useful for you to have a play to get some feedback on these
  -

**Comment by llwiggins** [comment]
Hi @stulacy - these features sound fab! I've checked out this branch and tried running the GUI from terminal as normal but seem to be running into errors. 

Terminal output below. Any ideas what could be causing this? It runs fine when I checkout main!

```
  You can now view your Streamlit app in your browser.

  URL: http://0.0.0.0:8501

*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow should only be instantiated on the main thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000193ed32ec __exceptionPreprocess + 176
	1   libobjc.A.dylib                     0x00000001939ba788 objc_exception_throw + 60
	2   CoreFoundation                      0x0000000193ef8268 _CFBundleGetValueForInfoKey + 0
	3   AppKit                              0x00000001976cd348 -[NSWindow _initContent:styleMask:backing:defer:contentView:] + 188
	4   AppKit                              0x00000001976cd280 -[NSWindow initWithContentRect:styleMask:backing:defer:] + 48
	5   libtk8.6.dylib                      0x000000033e429ba4 TkMacOSXMakeRealWindowExist + 564
	6   libtk8.6.dylib                      0x000000033e4297f0 TkWmMapWindow + 76
	7   libtk8.6.dylib                      0x000000033e386588 MapFrame + 76
	8   libtcl8.6.dylib                     0x000000033e7531a0 TclServiceIdle + 88
	9   libtcl8.6.dylib                     0x000000033e73309c Tcl_DoOneEvent + 120
	10  libtk8.6.dylib                      0x000000033e41a9c8 TkpInit + 800
	11  libtk8.6.dylib                      0x000000033e37ef20 Initialize + 2500
	12  _tkinter.cpython-312-darwin.so      0x0000000333c0a160 Tkapp_New + 876
	13  _tkinter.cpython-312-darwin.so      0x0000000333c09c1c _tkinter_create + 648
	14  python3.12                          0x00000001046342f8 cfunction_vectorcall_FASTCALL + 248
	15  python3.12                          0x00000001045d7f1c PyObject_Vectorcall + 88
	16  python3.12                          0x00000001046f198c _PyEval_EvalFrameDefault + 51644
	17  python3.12                          0x00000001045d7844 _PyObject_FastCallDictTstate + 324
	18  python3.12                          0x000000010465af38 slot_tp_init + 320
	19  python3.12                          0x00000001046532d8 type_call + 496
	20  python3.12                          0x00000001045d73bc _PyObject_MakeTpCall + 316
	21  python3.12                          0x00000001046f198c _PyEval_EvalFrameDefault + 51644
	22  python3.12                          0x00000001046e405c PyEval_EvalCode + 300
	23  python3.12                          0x00000001046e0c78 builtin_exec + 1412
	24  python3.12                          0x000000010463419c cfunction_vectorcall_FASTCALL_KEYWORDS + 144
	25  python3.12                          0x00000001045d7f1c PyObject_Vectorcall + 88
	26  python3.12                          0x00000001046f198c _PyEval_EvalFrameDefault + 51644
	27  python3.12                          0x00000001045db608 method_vectorcall + 372
	28  python3.12                          0x00000001046f385c _PyEval_EvalFrameDefault + 59532
	29  python3.12                          0x00000001045db608 method_vectorcall + 372
	30  python3.12                          0x00000001047f7e34 thread_run + 236
	31  python3.12                          0x0000000104773e1c pythread_wrapper + 48
	32  libsystem_pthread.dylib             0x0000000193d7ef94 _pthread_start + 136
	33  libsystem_pthread.dylib             0x0000000193d79d34 thread_start + 8
)
libc++abi: terminating due to uncaught exception of type NSException
zsh: abort      streamlit run GUI.py --server.port=8501 --server.address=0.0.0.0
```

**Comment by stulacy** [comment]
Ah that's really useful. Basically I've added in a folder selector to allow users to interactively choose the image directory rather than having to type it in to the text box. I've used a library called `tkinter` for this as this functionality isn't in Streamlit. `tkinter` comes with Python so I expected it to work the same on all operating systems, but it looks like it's doing something different on your Mac. I'll have a read about to see what the issue could be.

**Comment by stulacy** [comment]
Just to confirm, the above `tkinter` bug seems to be limited to MacOS. It runs fine on Windows and Linux

**Comment by stulacy** [comment]
I've confirmed that the above bug is only caused on Macs running on Apple silicon. I've created a fallback to use Laura's original textbox on these devices, on all others the file dialog window pops up.

@llwiggins Try it now, it should work although you won't be able to use the folder browser and will have to input the path directly as before

**Comment by llwiggins** [comment]
Thanks for your help with this @stulacy, it's now working on my Mac and your new changes look great! Happy for these to be merged to main 👍
